Server/data/stage.py

```diff
@@ -137,49 +167,98 @@ def __init__(self, verify=False, proxy_aware=True, ssl_versions=SecurityProtocol
+        ServicePointManager.Expect100Continue = False
 
     def post(self, url, payload=None):
         r = WebRequest.Create(url)
+        r.ServicePoint.ConnectionLimit = 500
+        r.Timeout = 30000
         r.ContentType = "application/octet-stream"
         r.Method = "POST"
         if self.proxy_aware:
             r.Proxy = WebRequest.GetSystemWebProxy()
             r.Proxy.Credentials = CredentialCache.DefaultCredentials
-
+       
         if len(payload):
             if type(payload) != Array[Byte]:
                 data = Encoding.UTF8.GetBytes(payload)
             else:
                 data = payload
-
             r.ContentLength = data.Length
-            requestStream = r.GetRequestStream()
-            requestStream.Write(data, 0, data.Length)
-            requestStream.Close()
-
-        return Response(r.GetResponse())
+            with r.GetRequestStream() as requestStream:
+                requestStream.Write(data, 0, data.Length)
+                requestStream.Close()
+        
+        return Response(r)
+        
 
     def get(self, url):
         r = WebRequest.Create(url)
+        r.ServicePoint.ConnectionLimit = 500
+        r.Timeout = 30000
         r.ContentType = "application/octet-stream"
         r.Method = "GET"
         if self.proxy_aware:
             r.Proxy = WebRequest.GetSystemWebProxy()
             r.Proxy.Credentials = CredentialCache.DefaultCredentials
-
-        return Response(r.GetResponse())
-
+        
+        return Response(r)
+        
 
 class Crypto(object):
     def __init__(self):
-        self.asymAlgo = AsymmetricAlgorithm.Create("ECDiffieHellmanCng")
-        self.public_key = self.asymAlgo.PublicKey.ToXmlString()
-        self.server_pubkey = None
-        self.derived_key = None
+        x9EC = SecNamedCurves.GetByName("secp521r1")
+        self.aliceKeyPair = self.GenerateKeyPair(ECDomainParameters(x9EC.Curve, x9EC.G, x9EC.N, x9EC.H, x9EC.GetSeed()))
+        self.bobPublicKey = self.GetBobPublicKey(URL, x9EC, self.aliceKeyPair.Public)
+
+        self.derived_key = self.derive_key(self.bobPublicKey, self.aliceKeyPair.Private)
+
+    def GenerateKeyPair(self, ecDomain):
+        g = GeneratorUtilities.GetKeyPairGenerator("ECDH")
+        g.Init(ECKeyGenerationParameters(ecDomain, SecureRandom()))
 
-    def derive_key(self, pubkey_xml):
-        self.server_pubkey = self.asymAlgo.PublicKey.FromXmlString(pubkey_xml)
-        self.derived_key = self.asymAlgo.DeriveKeyMaterial(self.server_pubkey)
+        return g.GenerateKeyPair()
+
+    def GetBobPublicKey(self, url, x9EC, alicePublicKey):
+        requests = Requests()
+
+        alice_x = str(alicePublicKey.Q.Normalize().AffineXCoord)
+        alice_y = str(alicePublicKey.Q.Normalize().AffineYCoord)
+        json = "{'x': \"" + alice_x + "\",'y': \"" + alice_y +"\"}"
+
+        response = requests.post(url, Encoding.UTF8.GetBytes(json)).text
+        response = response.replace("\"", "'")
+
+        r = Regex("': (.+?) ")
```

Comment:
I personally hate Regexes, if we could find a way to do without them i would be very happy

Reply:
Solved.